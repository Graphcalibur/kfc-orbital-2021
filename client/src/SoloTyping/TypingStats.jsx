import React from 'react';

const TypingStats = (props) => {
    /* Show no stats if typing test hasn't ended */
    if (!props.ended) return <></>;

    const { code, typed_wrong } = props;
    let code_length = 0;

    for (let i = 0; i < code.length; i++) {
      code_length += code[i].trim().length;
    }

    const accuracy =
      Math.round((code_length / (typed_wrong + code_length)) * 1000) / 10;
    const wpm = Math.round(code_length / 5 / (props.elapsed_time / 60000));

    return (
      <p className="text mt-3">
        Accuracy: {accuracy}% <br />
        Speed: {wpm} WPM
      </p>
    );
}

export default TypingStats;