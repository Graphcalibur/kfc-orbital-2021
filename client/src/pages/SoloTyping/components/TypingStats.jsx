import React from 'react';
import { Container, Button } from "react-bootstrap";

const TypingStats = (props) => {
    /* Show no stats if typing test hasn't ended */
    if (!props.ended) return <></>;

    const { code, typed_wrong } = props;
    let code_length = 0;

    for (let i = 0; i < code.length; i++) {
      code_length += code[i].trim().length; /* Don't count starting whitespace */
    }

    /* Accuracy = (# of chars in code / # of chars typed including wrong) * 100
       Formula does * 1000 / 10 so that it's accurate to the first decimal place */
    const accuracy =
      Math.round((code_length / (typed_wrong + code_length)) * 1000) / 10;

    /* WPM = (# of chars in code / 5) / time in minutes */
    const wpm = Math.round(code_length / 5 / (props.elapsed_time / 60000));

    return (
      <Container fluid="sm" className="shadow p-3 gap-3 mt-3 box">
        <h3 className="text">Your Typing Stats:</h3>

        <p className="text">
          Accuracy: {accuracy}% <br />
          Speed: {wpm} WPM
        </p>

        <Button onClick={props.reset} variant="primary" className="me-4">
          Try Again
        </Button>

        <Button
          onClick={() => {props.getCode(); props.reset();}}
          variant="primary" className="me-4"
        >
          New Practice
        </Button>
      </Container>
    );
}

export default TypingStats;