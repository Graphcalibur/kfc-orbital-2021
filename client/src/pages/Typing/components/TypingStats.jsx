import React from "react";
import { Container, Button } from "react-bootstrap";
import { useHistory } from "react-router-dom";

// TODO: Show "Log in to update your stats" if user is not logged in

const TypingStats = (props) => {
  /* Show no stats if typing test hasn't ended */
  if (!props.ended) return <span></span>;

  const { accuracy, wpm } = props;
  const history = useHistory();

  const getButtons = () => {
    if (props.is_solo) {
      return (
        <span>
          <Button
            onClick={props.reset}
            variant="outline-primary"
            className="me-4"
          >
            Try Again
          </Button>

          <Button
            onClick={() => {
              props.getCode();
              props.reset();
            }}
            variant="outline-primary"
            className="me-4"
          >
            New Practice
          </Button>
        </span>
      );
    } else {
      return (
        <span>
          <Button
            onClick={() => props.back_to_waiting(history)}
            variant="outline-primary"
            className="me-4"
          >
            Back to Waiting Room
          </Button>

          <Button
            onClick={() => history.push(`/rooms`)}
            variant="outline-primary"
          >
            Back to Room List
          </Button>
        </span>
      );
    }
  };

  return (
    <Container fluid="sm" className="shadow p-3 gap-3 mt-3 box">
      <h3 className="text">Your Typing Stats:</h3>

      <p className="text" id="stats">
        Accuracy: {accuracy}% <br />
        Speed: {wpm} WPM
      </p>

      {getButtons()}
    </Container>
  );
};

export default TypingStats;
