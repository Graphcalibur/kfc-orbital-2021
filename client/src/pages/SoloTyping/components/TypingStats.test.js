import React from "react";
import TypingStats from "./TypingStats";
import { create } from "react-test-renderer";

describe("Testing TypingStats component in page SoloTyping", () => {
  test("Test 1: Not ended", () => {
    const tree = create(<TypingStats ended={false} accuracy={100} wpm={42} />);
    expect(tree).toMatchSnapshot();
  });
  test("Test 2: Ended", () => {
    const tree = create(<TypingStats ended={false} accuracy={80} wpm={64.4} />);
    expect(tree).toMatchSnapshot();
  });
  test("Test 3: Ended", () => {
    const tree = create(<TypingStats ended={false} accuracy={92.3} wpm={71} />);
    expect(tree).toMatchSnapshot();
  });
});
