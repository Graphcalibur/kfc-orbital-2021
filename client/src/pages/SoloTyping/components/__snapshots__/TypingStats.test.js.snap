// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TypingStats component in page SoloTyping Test 1: Not ended 1`] = `<span />`;

exports[`Testing TypingStats component in page SoloTyping Test 2: Ended 1`] = `<span />`;

exports[`Testing TypingStats component in page SoloTyping Test 3: Ended 1`] = `<span />`;
